File: ocaml/plugins/doppler.ml
(* The `doppler` provider kind, as a voxgig/plugin definition.

   The Doppler API over HTTPS.

   A port of typescript/plugins/doppler.ts, which is canonical. *)

open Secret
open Provider
open Httpjson

(* Doppler.

   The whole config is downloaded ONCE - Doppler's own bulk endpoint - and
   answered from memory, like a remote .env: `api.token` is the `API_TOKEN`
   entry. A service token is config-scoped, so project and config are only
   needed with broader tokens. A failed load caches nothing, so it retries.

   The `prefix` option is deliberately not consulted by this kind. *)
let doppler_provider (token : string) (project : string) (config : string) (addr : string) :
    provider =
  let values = ref None in

  let load () =
    match !values with
    | Some loaded -> loaded
    | None ->
      let useaddr = trimslash (firstof [ addr; "https://api.doppler.com" ]) in
      checkaddr useaddr;

      let url = ref (useaddr ^ "/v3/configs/config/secrets/download?format=json") in
      if "" <> project then url := !url ^ "&project=" ^ Sigv4.uriescape project;
      if "" <> config then url := !url ^ "&config=" ^ Sigv4.uriescape config;

      let res = fetchjson ~headers:[ ("authorization", "Bearer " ^ token) ] "GET" !url in
      let body = Json.oasobj res.jbody in

      if 200 <> res.status || None = body then
        fail ("sekreto: doppler error: " ^ string_of_int res.status);

      let loaded =
        List.filter_map
          (fun (key, value) -> match Json.text value with Some text -> Some (key, text) | None -> None)
          (Option.get body)
      in

      values := Some loaded;
      loaded
  in

  {
    lookup = (fun name -> List.assoc_opt (envkey name) (load ()));
    describe =
      (fun () -> if "" = project then "doppler" else "doppler:" ^ project ^ "/" ^ config);
  }

let plugin () : Defs.definition =
  providerplugin "doppler" (fun spec ->
      doppler_provider spec.token spec.project spec.config spec.addr)

File: elixir/plugins/doppler.ex
# Doppler, as a sekreto plugin.
#
# A port of typescript/plugins/doppler.ts, which is canonical.

defmodule Sekreto.Plugins.Doppler do
  @moduledoc """
  The `doppler` provider kind.

  A plugin module: nothing under `src/` names it, and a chain that does not
  name this kind never loads it. See docs/design/plugin-providers.md.
  """

  alias Sekreto.Cell
  alias Sekreto.Error
  alias Sekreto.Json
  alias Sekreto.Plugins.Httpjson
  alias Sekreto.Plugins.Sigv4
  alias Sekreto.ProviderSpec
  alias Sekreto.Providers

  @doc """
  The `doppler` provider kind, as a voxgig/plugin definition.

  The calling project passes it to the constructor, and only then can a
  chain name the kind:

      Sekreto.new(chain, plugins: [Sekreto.Plugins.Doppler.doppler()])
  """
  def doppler, do: Providers.providerplugin("doppler", &fromspec/1)

  @doc """
  Doppler.

  The whole config is downloaded once - Doppler's own bulk endpoint - and
  answered from memory, like a remote .env: `api.token` is the `API_TOKEN`
  entry. A service token is config-scoped, so project and config are only
  needed with broader tokens.

  The `prefix` option is not consulted by this kind.
  """
  def doppler(token, project, config, addr) do
    cell = Cell.new(:unloaded)

    load = fn ->
      useaddr = Httpjson.trimslash(Providers.first([addr, "https://api.doppler.com"]))
      Providers.checkaddr(useaddr)

      url =
        useaddr <>
          "/v3/configs/config/secrets/download?format=json" <>
          if("" == project, do: "", else: "&project=" <> Sigv4.uriescape(project)) <>
          if("" == config, do: "", else: "&config=" <> Sigv4.uriescape(config))

      res = Httpjson.fetchjson("GET", url, [{"authorization", "Bearer " <> token}])
      body = Json.asobj(res.body)

      if 200 != res.status or :none == body do
        raise Error, message: "sekreto: doppler error: " <> Httpjson.tostr(res.status)
      end

      Enum.reduce(body, [], fn {key, value}, out ->
        case Json.text(value) do
          :none -> out
          text -> Sekreto.pairput(out, key, text)
        end
      end)
    end

    %{
      lookup: fn name ->
        values =
          case Cell.get(cell) do
            :unloaded ->
              # A failed load caches nothing, so the next lookup retries.
              loaded = load.()
              Cell.put(cell, loaded)
              loaded

            loaded ->
              loaded
          end

        Sekreto.pairget(values, Sekreto.envkey(name))
      end,
      describe: fn ->
        if "" == project, do: "doppler", else: "doppler:" <> project <> "/" <> config
      end
    }
  end

  defp fromspec(%ProviderSpec{} = spec),
    do: doppler(spec.token, spec.project, spec.config, spec.addr)
end
